core/src/main/scala/org/apache/spark/deploy/history/FsHistoryProvider.scala

```diff
@@ -187,47 +201,78 @@ private[history] class FsHistoryProvider(conf: SparkConf) extends ApplicationHis
+            itor.next()
+          }
+        }
+      }
 
       lastModifiedTime = newLastModifiedTime
+    } catch {
+      case e: Exception => logError("Exception in checking for event log updates", e)
+    }
+  }
 
-      // When there are new logs, merge the new list with the existing one, maintaining
-      // the expected ordering (descending end time). Maintaining the order is important
-      // to avoid having to sort the list every time there is a request for the log list.
-      if (!logInfos.isEmpty) {
-        val newApps = new mutable.LinkedHashMap[String, FsApplicationHistoryInfo]()
-        def addIfAbsent(info: FsApplicationHistoryInfo) = {
-          if (!newApps.contains(info.id) ||
-              newApps(info.id).logPath.endsWith(EventLoggingListener.IN_PROGRESS) &&
-              !info.logPath.endsWith(EventLoggingListener.IN_PROGRESS)) {
-            newApps += (info.id -> info)
-          }
-        }
+  /**
+   * Fetch and Parse the log files
+   */
+  private def replay(logs: Iterator[FileStatus]) {
+    def addIfAbsent(newApps: mutable.LinkedHashMap[String, FsApplicationHistoryInfo],
+                    info: FsApplicationHistoryInfo) {
+      if (!newApps.contains(info.id) ||
+        newApps(info.id).logPath.endsWith(EventLoggingListener.IN_PROGRESS) &&
+          !info.logPath.endsWith(EventLoggingListener.IN_PROGRESS)) {
+        newApps += (info.id -> info)
+      }
+    }
 
-        val newIterator = logInfos.iterator.buffered
-        val oldIterator = applications.values.iterator.buffered
-        while (newIterator.hasNext && oldIterator.hasNext) {
-          if (compareAppInfo(newIterator.head, oldIterator.head)) {
-            addIfAbsent(newIterator.next)
-          } else {
-            addIfAbsent(oldIterator.next)
-          }
-        }
-        newIterator.foreach(addIfAbsent)
-        oldIterator.foreach(addIfAbsent)
+    def mergeApps(list: Seq[FsApplicationHistoryInfo]) {
+      val newApps = new mutable.LinkedHashMap[String, FsApplicationHistoryInfo]()
 
-        applications = newApps
+      val newIterator = list.iterator.buffered
+      val oldIterator = applications.values.iterator.buffered
+      while (newIterator.hasNext && oldIterator.hasNext) {
+        if (compareAppInfo(newIterator.head, oldIterator.head)) {
+          addIfAbsent(newApps, newIterator.next())
+        } else {
+          addIfAbsent(newApps, oldIterator.next())
+        }
       }
-    } catch {
-      case e: Exception => logError("Exception in checking for event log updates", e)
+      newIterator.foreach(addIfAbsent(newApps, _))
+      oldIterator.foreach(addIfAbsent(newApps, _))
+
+      applications = newApps
     }
+
+    val bus = new ReplayListenerBus()
+    val newApps = logs.flatMap { fileStatus =>
+      try {
+        val res = replay(fileStatus, bus)
+        logInfo("load log " + res.logPath + " successfully")
+        Some(res)
+      } catch {
+        case e: Exception => None
```

Comment:
should we log some error here? Otherwise we silently fail.
```
logError(s"Exception encountered when attempting to load application log ${fileStatus.getPath}")
```